README
NAME
    Finance::Currency::Convert::WebserviceX - Lightweight currency
    conversion using WebserviceX.NET

SYNOPSIS
        use Finance::Currency::Convert::WebserviceX;
        my $cc = Finance::Currency::Convert::WebserviceX->new;
        my $result = $cc->convert(1.95, 'USD', 'JPY');

DESCRIPTION
    This is a lightweight module to do currency conversion using the
    Currency Converter web service at http://www.webservicex.net/.

    The motivation for this module was many fold. First,
    Finance::Currency::Convert with Finance::Quote was a little too bulky
    for my needs, esp the need to download or maintain conversion tables.
    Finance::Currency::Convert::Yahoo seemed to be based on screen scraping.
    Way to fragile for my taste. Finance::Currency::Convert::XE has usage
    restrictions from XE.com. [No offense intended to any of the authors
    above]

CONSTRUCTOR
    You know the routine. "new" is your friend.

        use Finance::Currency::Convert::WebserviceX;
        my $cc = Finance::Currency::Convert::WebserviceX->new;

METHODS
  convert($value, $from, $to)
    Converts a number value from one currency to another and returns the
    result.

        my $result = $cc->convert(1.95, 'USD', 'JPY');

    If an error occurs, no value is given, or the from/to aren't 3 letter
    currency codes, "convert" returns "undef".

    For now, you can access the request response after calling C>convert>:

        my $response = $self->{'response'};

    This returns a HTTP::Response object that can be used to inspect any
    remote web service errors. $self->response{'request'} is reset at the
    beginning of every call to "convert" and returns "undef" otherwise.

    value
        The number or price to be converted.

    from
        The three letter ISO currency code for the currency amount specified
        in "value". See Locale::Currency for the available currency codes.

    to  The three letter ISO currency code for the currency you want the
        "value" to be converted to. See Locale::Currency for the available
        currency codes.

SEE ALSO
    Locale::Currency, Finance::Currency::Format

AUTHOR
        Christopher H. Laco
        CPAN ID: CLACO
        claco@chrislaco.com
        http://today.icantfocus.com/blog/


Changes
$Id: /local/CPAN/Finance-Currency-Convert-WebserviceX/Changes 1665 2008-06-04T00:20:50.790988Z claco  $
Revision history for Perl module Finance::Currency::Convert::WebserviceX

0.07000 Tue June 3 19:36:46 2008
    - Added caching of webservice results; default of 300 seconds per from/to
        pair of currency codes

0.06002 Sun Jan 20 12:51:43 2007
    - Updated broken META.yml
    - Added convert mock tests. Live converts use TEST_AUTHOR

0.06001 Fri July 13 21:53:47 2007
    - Fixed MANIFEST.SKIP for missing README

0.06000 Sat June 30 17:54:56 2007
    - Converted to Module::Install
    - Added more culted tests

0.05 Mon Mar 21 18:56:24 2005
    - Added Test::Strict warnings_ok tests

0.04 Mon Mar 14 19:00:23 2005
    - Added Test::Strict tests

0.03 Thur Mar 10 20:19:27 2005
    - Fixed bug where converting from XXX to XXX yielded a price of 0
        since the rate returned was also 0

0.02 Web Mar 09 22:45:13 2005
    - Added missing convert example to pod

0.01 Wed Mar 09 20:21:35 2005
    - Initial release


lib/Finance/Currency/Convert/WebserviceX.pm
# $Id: /local/CPAN/Finance-Currency-Convert-WebserviceX/lib/Finance/Currency/Convert/WebserviceX.pm 1665 2008-06-04T00:20:50.790988Z claco  $
package Finance::Currency::Convert::WebserviceX;
use strict;
use warnings;
use vars qw($VERSION);
use LWP::UserAgent ();
use Memoize::Expire ();

$VERSION = '0.07000';

my $expires = tie my %cache => 'Memoize::Expire', LIFETIME => 300;

sub new {
    my $class = shift;
    my $self = bless{response => undef}, ref $class || $class;

    my $ua = LWP::UserAgent->new;
    $ua->timeout(10);
    $ua->env_proxy;

    $self->{'ua'} = $ua;

    return $self;
};

sub cache {
    return \%cache;
}

sub convert {
    my $self = shift;
    my $value = shift || '';
    my $from = shift || '';
    my $to = shift || '';
    my $key = "$from-$to";

    $from = uc($from);
    $to   = uc($to);

    return unless length $value && $from =~ /^[A-Z]{3}$/ && $to =~ /^[A-Z]{3}$/;

    if ($from eq $to) {
        return $value;
    }

    if (exists $cache{$key}) {
        return $cache{$key};
    }

    my $uri = sprintf('http://www.webservicex.net/CurrencyConvertor.asmx/ConversionRate?FromCurrency=%s&ToCurrency=%s', $from, $to);

    eval {
        $self->{'response'} = undef;
        $self->{'response'} = $self->{'ua'}->get($uri);
    };

    return if $@;

    if (!$self->{'response'}->is_success) {
        return;
    } else {
        if (($self->{'response'}->content || '') =~ /<double.*>(.*)<\/double>/i) {
            my $rate = $value*($1 || 1);
            $cache{$key} = $rate;
            return $rate;
        } else {
            return;
        };
    };
};

1;
__END__

=head1 NAME

Finance::Currency::Convert::WebserviceX - Lightweight currency conversion using WebserviceX.NET

=head1 SYNOPSIS

    use Finance::Currency::Convert::WebserviceX;
    my $cc = Finance::Currency::Convert::WebserviceX->new;
    my $result = $cc->convert(1.95, 'USD', 'JPY');

=head1 DESCRIPTION

This is a lightweight module to do currency conversion using the Currency
Converter web service at http://www.webservicex.net/.

The motivation for this module was many fold. First,
L<Finance::Currency::Convert> with L<Finance::Quote> was a little too bulky for
my needs, esp the need to download or maintain conversion tables.
L<Finance::Currency::Convert::Yahoo> seemed to be based on screen scraping. Way
to fragile for my taste. L<Finance::Currency::Convert::XE> has usage
restrictions from XE.com. [No offense intended to any of the authors above]

=head1 CONSTRUCTOR

You know the routine. C<new> is your friend.

    use Finance::Currency::Convert::WebserviceX;
    my $cc = Finance::Currency::Convert::WebserviceX->new;


=head1 METHODS

=head2 convert($value, $from, $to)

Converts a number value from one currency to another and returns the result.

    my $result = $cc->convert(1.95, 'USD', 'JPY');

If an error occurs, no value is given, or the from/to aren't 3 letter currency codes,
C<convert> returns C<undef>.

For now, you can access the request response after calling C>convert>:

    my $response = $self->{'response'};

This returns a L<HTTP::Response> object that can be used to inspect any remote web
service errors. $self->response{'request'} is reset at the beginning of every call
to C<convert> and returns C<undef> otherwise.

=over

=item value

The number or price to be converted.

=item from

The three letter ISO currency code for the currency amount
specified in C<value>. See L<Locale::Currency> for the available
currency codes.

=item to

The three letter ISO currency code for the currency you want the
C<value> to be converted to. See L<Locale::Currency> for the available
currency codes.

=back

=head2 cache

Gets the reference to the cache hash.

=head1 SEE ALSO

L<Locale::Currency>, L<Finance::Currency::Format>, L<Memoize::Expire>

=head1 AUTHOR

    Christopher H. Laco
    CPAN ID: CLACO
    claco@chrislaco.com
    http://today.icantfocus.com/blog/
